### PR TITLE
Support for external embeddings with variable size

### PR DESCRIPTION
**Pull request by Edresson**
Status: In Progress

 *  [x] Tacotron 1 support for external embeddings with variable size
 * [x] Ipython notebook for extract embeddings from speaker_encoder ( in the expected format)
 *  [x]  Support BRSpeech Potuguese dataset
 * [x]  Ipython notebook for extracting embeddigns from the [Speech2Phone](https://arxiv.org/abs/2002.11213) system
* [x] Tacotron 2 support for external embeddings with variable size
*  [ ] Experiment 1:  Concatenate the Speaker Embedding on the StopNet (in a multi speaker scenario the big problem is in the convergence of the stop token loss, this should help)
* [ ]  Experiment 2: Support for embeddings by samples and not by speaker (each speaker sample will have an embedding), this should have better results for voice cloning. (I do not see how to implement this while maintaining the current implementation that uses nn.Embedding, I will test in a branch itself in case of positive results we can think of how to add this !)

### CODE REVIEW

**Comment by Edresson** [comment]
@erogol  I followed the same approach as tacotron 1, however as the encoder of tacotron 2 does not have PreNet, I am not passing the speaker embedding to the encoder in this way I am just feeding the decoder.

I think that passing the embedding to the encoder would also make the decoder_dim very large (decoder_dim = 512 + speaker_embedding_dim *2), since the encoder in tacotron 2 returns a tensor with dimension (512 + speaker_embedding_dim).

What do you think ? is it ideal?

**Comment by erogol** [comment]
> 
> 
> @erogol Tacotron2 does not have speaker_project_mel in the part of the embeddings of speakers.
> 
> Support for tacatron2 multi-speaker training is quite different from tacotron1. In my tests as it is, it is not possible to add embeddings that are not 512 in size. I believe it is very desirable to use embeddings with different sizes, in order to support other speaker encoders.
> 
> Do you suggest following the same approach as tacotron 1 and adding speaker_project_mel?

I think tacotron approach is the most flexible one with less code update.

**Comment by erogol** [comment]
> 
> 
> @erogol I followed the same approach as Tacotron 1, however as the encoder of Tacotron 2 does not have PreNet, I am not passing the speaker embedding to the encoder in this way I am just feeding the decoder.
> 
> I think that passing the embedding to the encoder would also make the decoder_dim very large (decoder_dim = 512 + speaker_embedding_dim *2), since the encoder in Tacotron 2 returns a tensor with dimension (512 + speaker_embedding_dim).
> 
> What do you think ? is it ideal?

The way I connect the inputs are taken from the latest DeepVoice paper. But it needs more experimentation to tell which approach is better. Since I was lacking these experiments, I preferred to connect speaker embeddings basically to every module in the model. Kind of a lazy approach. But I am open to any updates if they work any better.

**Comment by erogol** [comment]
Please also consider adding a some tests since it's a big update, and we don't like to break anything hidden.

**Comment by Edresson** [comment]
> 
> 
> Please also consider adding a some tests since it's a big update, and we don't like to break anything hidden. Thx for this great contribution.



> 
> 
> Please also consider adding a some tests since it's a big update, and we don't like to break anything hidden. Thx for this great contribution.

OK, I'm going to do the unit tests.

Now I'm testing whether concatenating the speaker's embeddings on StopNet helps the stop token loss converge in the multi-speaker scenario. In my tests this is the main problem for some speakers, StopNet works very well but for others it doesn't.

**Comment by erogol** [comment]
I'd worry about the stopnet at last. If even it does not work, you can use the attention alignment to stop. I'd suggest you to worry about the quality first.

**Comment by Edresson** [comment]
> 
> 
> I'd worry about the stopnet at last. If even it does not work, you can use the attention alignment to stop. I'd suggest you to worry about the quality first.
I agree,

I have the problem of an speaker pronouncing the entire sentence perfectly (StopNet stopping). And other speakers not saying the entire sentence (and it is  Decoder stopped with max_decoder_steps). From what seems to me an average of the embeddings of each sample, it does not seem enough for the model to learn all the characteristics, given that the dataset I am using is not a dataset made only of samples of professional speakers (professional speakers generally maintain the same speech rate , that is, a more constant speech rate). When you have non-professional speakers in the dataset I believe that the embeddings average is not enough. Because the speed of speech changes a lot from sample to another.

I'm running an experiment where each sample has its own specific embedding (an speaker has several different embeddings), so I believe it will be easier for the model to learn the speech rate correctly.

**Comment by erogol** [comment]
you can visualize the embeddings. I believe if the embeddings are too close for the same speaker against the other speakers, I'd say average embedding vector should be enough. However, from my limited experiments with multi-speaker setting, model needs to be trained quite longer than a single speaker model. How long did you trained so far?

**Comment by Edresson** [comment]
> 
> 
> you can visualize the embeddings. I believe if the embeddings are too close for the same speaker against the other speakers, I'd say average embedding vector should be enough. However, from my limited experiments with multi-speaker setting, model needs to be trained quite longer than a single speaker model. How long did you trained so far?



The last train I stopped came to approximately 190k steps. I stopped because the loss of validation and training started to increase a lot. And after I synthesized with these checkpoints and realized that only one of the announcers pronounced a certain sentence the others stopped before the sentence ended.

**Comment by erogol** [comment]
My model worked good for female voices but not that much for male voices. (I tried LibriTTS once with not many experiments). So maybe providing the gender as an explicit attribute might also help. I also see papers learning different BatchNorm stats for different speakers.

**Comment by Edresson** [comment]
> My model worked good for female voices but not that much for male voices. (I tried LibriTTS once with not many experiments). So maybe providing the gender as an explicit attribute might also help. I also see papers learning different BatchNorm stats for different speakers.

Apparently the model that receives one embedding per sample, that is, several embeddings per speaker, since receiving only one embedding, is doing better with less training steps.

Follow the test samples : 

[Embedding per sample ( one speaker have various embeddings ) with 84915 steps ](https://drive.google.com/file/d/1nEvKerCEsRXrzjLRlt1Q_0oKXgNmR_8U/view?usp=sharing)

[Actual implementation (one speaker have one embedding) with 90000 steps ](https://drive.google.com/file/d/1eEOaht6_N7YQwnSUlRard-uVmRm-wBNW/view?usp=sharing)

Realize that the model trained with several embeddings sounds more natural. However, the model has a problem with pronunciation in one of the 20 phonetically balanced phrases, but I believe that with more training this will be corrected.

The models are still training this quality of synthesization is only in professional speakers (who maintain a good intonation and rhythm of speech), the other speakers do not pronounce all the words of the phrases and the synthesis is for max_decoder_steps.


@erogol  The decoder receives the embedding after it passes through the speaker_project_mel (embedding -> FC -> decoder), while the Encoder receives the embedding directly (Embedding -> Encoder). Have you tried to pass this same embedding to the Encoder or create another layer for it?

What about the size of the speaker_project_mel (80) did you do any experiments with higher or lower values?

**Comment by erogol** [comment]
I don't remember that well but I don't think I've tried a projection layer in Encoder (if that's is what you mean). It is hard to guess how it'd work. It is better to just try :)

I've not tried different values for the mel-spectrogram. 80 is the standard. I don't think it'd cause a big perf difference. Also, computational complexity increases almost exponentially as you increase the size.

**Comment by erogol** [comment]
Btw samples with individual embedding vectors sound good. I'd suggest you to use the universal WaveRNN model to test if the results get any better. 

https://github.com/erogol/WaveRNN

It requires 16khz sampling rate. Hope it matches your dataset.

**Comment by Edresson** [comment]
> I don't remember that well but I don't think I've tried a projection layer in Encoder (if that's is what you mean). It is hard to guess how it'd work. It is better to just try :)
> 
> I've not tried different values for the mel-spectrogram. 80 is the standard. I don't think it'd cause a big perf difference. Also, computational complexity increases almost exponentially as you increase the size.

If the model does not show satisfactory results I will try this.




> Btw samples with individual embedding vectors sound good. I'd suggest you to use the universal WaveRNN model to test if the results get any better.
> 
> https://github.com/erogol/WaveRNN
> 
> It requires 16khz sampling rate. Hope it matches your dataset.

My dataset is at 22khz, but I have fine-tuned Universal WaveRNN for a Portuguese dataset that I can use.

At the moment I will continue using Griffin-Lim, as soon as the training is over I can test with Universal WaveRNN, thanks :).

**Comment by erogol** [comment]
Let me know how it'll go with WaveRNN.

**Comment by Edresson** [comment]
@erogol  I created a TacotronExternalEmbeddingsTrainTest class in the files test/test_tacotron_model.py and test/test_tacotron2_model.py. The class simulates training with an external embedding (randomly generated). Do you think this is enough?

**Comment by erogol** [comment]
it is better if you can check against an expected output. I don't know exactly how but it'd be better :). But now thinking about it

**Comment by Edresson** [comment]
The best results I got were implemented in PR #472 , so I will keep this PR closed!
